webapp/channels/src/utils/performance_telemetry/reporter.ts

```diff
@@ -0,0 +1,277 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import type {Store} from 'redux';
+import {onCLS, onFCP, onINP, onLCP, onTTFB} from 'web-vitals';
+import type {Metric} from 'web-vitals';
+
+import type {Client4} from '@mattermost/client';
+
+import {getConfig} from 'mattermost-redux/selectors/entities/general';
+import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
+
+import type {GlobalState} from 'types/store';
+
+import type {PerformanceLongTaskTiming} from './long_task';
+
+type PerformanceReportMeasure = {
+    metric: string;
+    value: number;
+    timestamp: number;
+}
+
+type PerformanceReport = {
+
+    // TODO confirm the version number that we want to start with
+    version: '0';
+
+    // TODO client ID?
+
+    platform: string;
+    user_agent: string;
```

Comment:
We can convert these to map to have some extensibility.

Reply:
I wanted to talk about that actually since I wanted the server to be responsible for determining what the user's browser/OS is instead of the web app. It already has a library for doing that which it uses for populating the Sessions table, and I trust that library better than the custom code we have on the web app for doing that. We can discuss that today

Reply:
As discussed, we're going to detect that on the client and then validate it on the server. That way, the client can detect the user agent once and hold onto the value while we prevent malicious clients from generating arbitrary time series on the server by sending arbitrary strings with metrics

Reply:
Added